core/src/main/scala/zio/config/ReadFunctions.scala

```diff
@@ -11,29 +11,51 @@ private[config] trait ReadFunctions {
     def loop[V1, B](
       configuration: ConfigDescriptor[K, V1, B],
       paths: Vector[K]
-    ): ZIO[Any, ReadErrorsVector[K, V1], ::[B]] =
+    ): ZIO[Any, Either[ReadErrorsVector[K, V1], ReadFunctions.MissingValuesInList[K, V1, B]], ::[B]] =
       configuration match {
         case ConfigDescriptor.Source(path, source: ConfigSource[K, V1], propertyType: PropertyType[V1, B]) =>
-          for {
-            value <- source.getConfigValue(paths :+ path)
-            result <- foreach(value.value) { eachValue =>
-                       ZIO.fromEither(
-                         propertyType
-                           .read(eachValue)
-                           .fold(
-                             r =>
-                               Left(
-                                 singleton(
-                                   ReadError
-                                     .parseError(paths :+ path, r.value, r.typeInfo): ReadError[Vector[K], V1]
-                                 )
-                               ),
-                             e => Right(e)
-                           )
-                       )
+          val results = for {
+            value <- source
+                      .getConfigValue(paths :+ path)
+                      .mapError(t => Left(t))
+
+            result <- foreach(value.value) {
+                       case Some(value) =>
+                         ZIO.fromEither(
+                           propertyType
+                             .read(value)
+                             .fold(
+                               r =>
+                                 Left(
+                                   Left(
+                                     singleton(
+                                       ReadError
+                                         .parseError(paths :+ path, r.value, r.typeInfo): ReadError[Vector[K], V1]
+                                     )
+                                   )
+                                 ),
+                               e => Right(Some(e))
+                             )
+                         )
+                       case None => ZIO.succeed(None: Option[B])
                      }
           } yield result
 
+          results.flatMap(
+            list =>
+              if (list.exists(_.isEmpty))
+                ZIO.fail(
+                  Right(
+                    ReadFunctions.MissingValuesInList[K, V1, B](list, paths)
+                  ): Either[ReadErrorsVector[K, V1], ReadFunctions.MissingValuesInList[K, V1, B]]
+                )
+              else
+                ZIO.succeed({
+                  val sList = list.flatMap(_.toList)
+                  ::(sList.head, sList.tail)
+                })
+          )
```

Comment:
Store indices and number of missing value as part of ReadError (in follow up PR)

Reply:
This will be a huge win when dealing with type safe config's list values having empty values in between